kotlin codege: IllegalStateException: No type argument found for 0!

```kotlin
@JsonClass(generateAdapter = true)
class TextAssetMetaData: AssetMetaData<TextAsset>()

class TextAsset: Asset<TextAsset>()

abstract class Asset<A: Asset<A>>
abstract class AssetMetaData<A: Asset<A>>
```
This crashes with 
```
[Kapt and Gradle exception]
Caused by: java.lang.IllegalStateException: No type argument found for 0!
        at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs$toTypeParamsResolver$typeParamResolver$1.invoke(KotlinPoetMetadataSpecs.kt:198)
        at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs$toTypeParamsResolver$typeParamResolver$1.invoke(KotlinPoetMetadataSpecs.kt)
        at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName(KmTypes.kt:84)
        at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName$default(KmTypes.kt:79)
        at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName(KmTypes.kt:62)
        at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName(KmTypes.kt:81)
        at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName$default(KmTypes.kt:79)
        at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeVariableName(KmTypes.kt:156)
        at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeParamsResolver(KotlinPoetMetadataSpecs.kt:201)
        at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeParamsResolver$default(KotlinPoetMetadataSpecs.kt:192)
        at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeSpec(KotlinPoetMetadataSpecs.kt:211)
        at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeSpec(KotlinPoetMetadataSpecs.kt:173)
        at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeSpec(KotlinPoetMetadataSpecs.kt:144)
        at com.squareup.moshi.kotlin.codegen.MetadataKt.declaredProperties$default(metadata.kt:206)
        at com.squareup.moshi.kotlin.codegen.MetadataKt.targetType(metadata.kt:184)
        at com.squareup.moshi.kotlin.codegen.JsonClassCodegenProcessor.adapterGenerator(JsonClassCodegenProcessor.kt:133)
        at com.squareup.moshi.kotlin.codegen.JsonClassCodegenProcessor.process(JsonClassCodegenProcessor.kt:107)
...
```
I am pretty sure this is a bug in kotlinpoet, but I'm not really sure how to repoduce it at their end. My testing shows that the bug is in connection with the circular generic definition on `AssetMetaData<A: Asset<A>>`. 

Tested with current SNAPSHOT.

Can you reproduce with Kotlinpoet-metadata directly? If so we can transfer
this issue to there

On Thu, Oct 17, 2019 at 2:12 AM Burkhard Mittelbach <
notifications@github.com> wrote:

> @JsonClass(generateAdapter = true)class TextAssetMetaData: AssetMetaData<TextAsset>()
> class TextAsset: Asset<TextAsset>()
> abstract class Asset<A: Asset<A>>abstract class AssetMetaData<A: Asset<A>>
>
> This crashes with
>
> [Kapt and Gradle exception]
> Caused by: java.lang.IllegalStateException: No type argument found for 0!
>         at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs$toTypeParamsResolver$typeParamResolver$1.invoke(KotlinPoetMetadataSpecs.kt:198)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs$toTypeParamsResolver$typeParamResolver$1.invoke(KotlinPoetMetadataSpecs.kt)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName(KmTypes.kt:84)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName$default(KmTypes.kt:79)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName(KmTypes.kt:62)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName(KmTypes.kt:81)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeName$default(KmTypes.kt:79)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.internal.KmTypesKt.toTypeVariableName(KmTypes.kt:156)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeParamsResolver(KotlinPoetMetadataSpecs.kt:201)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeParamsResolver$default(KotlinPoetMetadataSpecs.kt:192)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeSpec(KotlinPoetMetadataSpecs.kt:211)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeSpec(KotlinPoetMetadataSpecs.kt:173)
>         at com.squareup.moshi.kotlinpoet.metadata.specs.KotlinPoetMetadataSpecs.toTypeSpec(KotlinPoetMetadataSpecs.kt:144)
>         at com.squareup.moshi.kotlin.codegen.MetadataKt.declaredProperties$default(metadata.kt:206)
>         at com.squareup.moshi.kotlin.codegen.MetadataKt.targetType(metadata.kt:184)
>         at com.squareup.moshi.kotlin.codegen.JsonClassCodegenProcessor.adapterGenerator(JsonClassCodegenProcessor.kt:133)
>         at com.squareup.moshi.kotlin.codegen.JsonClassCodegenProcessor.process(JsonClassCodegenProcessor.kt:107)
> ...
>
> I am pretty sure this is a bug in kotlinpoet, but I'm not really sure how
> to repoduce it at their end. My testing shows that the bug is in connection
> with the circular generic definition on AssetMetaData<A: Asset<A>>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/square/moshi/issues/955?email_source=notifications&email_token=AAKMJPWXAD2J5L7U2BACQOLQO764VA5CNFSM4JBUWSUKYY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HSLMSAQ>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKMJPQZAZSPFJR3UXFKRXLQO764VANCNFSM4JBUWSUA>
> .
>

--- a/kotlin/codegen/pom.xml
+++ b/kotlin/codegen/pom.xml
@@ -218,10 +218,6 @@
                   <shadedPattern>com.squareup.moshi.kotlinpoet.classinspector</shadedPattern>
                 </relocation>
                 <relocation>
-                  <pattern>kotlinpoet.classinspector.elements.shaded</pattern>
-                  <shadedPattern>com.squareup.moshi.kotlinpoet.classinspector.elements.shaded</shadedPattern>
-                </relocation>
-                <relocation>
                   <pattern>kotlinx.metadata</pattern>
                   <shadedPattern>com.squareup.moshi.kotlinx.metadata</shadedPattern>
                 </relocation>

--- a/kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt
+++ b/kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt
@@ -12,6 +12,7 @@ import com.squareup.moshi.Types
 import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
 import com.squareup.moshi.kotlin.reflect.adapter
 import org.assertj.core.api.Assertions.assertThat
+import org.intellij.lang.annotations.Language
 import org.junit.Assert.fail
 import org.junit.Test
 import org.junit.runner.RunWith
@@ -273,13 +274,34 @@ class DualKotlinTest(useReflection: Boolean) {
 
     val consumer = InlineConsumer(InlineClass(23))
 
-    val expectedJson= """{"inline":{"i":23}}"""
+    @Language("JSON")
+    val expectedJson = """{"inline":{"i":23}}"""
     assertThat(adapter.toJson(consumer)).isEqualTo(expectedJson)
 
+    @Language("JSON")
     val testJson = """{"inline":{"i":42}}"""
     val result = adapter.fromJson(testJson)!!
     assertThat(result.inline.i).isEqualTo(42)
   }
+
+  // Regression test for https://github.com/square/moshi/issues/955
+  @Test fun backwardReferencingTypeVars() {
+    val adapter = moshi.adapter<TextAssetMetaData>()
+
+    @Language("JSON")
+    val testJson = """{"text":"text"}"""
+
+    assertThat(adapter.toJson(TextAssetMetaData("text"))).isEqualTo(testJson)
+
+    val result = adapter.fromJson(testJson)!!
+    assertThat(result.text).isEqualTo("text")
+  }
+
+  @JsonClass(generateAdapter = true)
+  class TextAssetMetaData(val text: String) : AssetMetaData<TextAsset>()
+  class TextAsset : Asset<TextAsset>()
+  abstract class Asset<A : Asset<A>>
+  abstract class AssetMetaData<A : Asset<A>>
 }
 
 // Has to be outside since inline classes are only allowed on top level

--- a/pom.xml
+++ b/pom.xml
@@ -36,8 +36,8 @@
     <incap.version>0.2</incap.version>
     <okio.version>1.16.0</okio.version>
     <okio2.version>2.1.0</okio2.version>
-    <kotlin.version>1.3.40</kotlin.version>
-    <kotlinpoet.version>1.4.0</kotlinpoet.version>
+    <kotlin.version>1.3.50</kotlin.version>
+    <kotlinpoet.version>1.4.1</kotlinpoet.version>
     <kotlinx-metadata.version>0.1.0</kotlinx-metadata.version>
     <maven-assembly.version>3.1.0</maven-assembly.version>
 

